solidity/contracts/BancorNetwork.sol

```diff
@@ -450,6 +405,191 @@ contract BancorNetwork is IBancorNetwork, TokenHolder, ContractRegistryClient, F
         return claimAndConvertFor2(_path, _amount, _minReturn, msg.sender, _affiliateAccount, _affiliateFee);
     }
 
+    /**
+      * @dev validates msg.value and prepares the conversion source token for the conversion
+      * 
+      * @param _sourceToken source token of the first conversion step
+      * @param _smartToken  smart token of the first conversion step
+      * @param _amount      amount to convert from, in the source token
+    */
+    function handleSourceToken(IERC20Token _sourceToken, ISmartToken _smartToken, uint256 _amount) private {
+        bool isSourceETH = etherTokens[_sourceToken];
+
+        // validate msg.value
+        if (msg.value > 0)
+            require(msg.value == _amount && isSourceETH);
+        else
+            require(msg.value == 0);
+
+        IBancorConverter firstConverter = IBancorConverter(_smartToken.owner());

```

Comment:
We don't need to execute this in all scenarios.
You can just embed `IBancorConverter(_smartToken.owner())` in each one of the two cases where it is actually needed.

Reply:
It's used in all the cases since this relies on it -
`bool isNewerConverter = isV27OrHigherConverter(firstConverter);`